### PR TITLE
codeintel: Add janitor process to api-server

### PR DESCRIPTION
**Pull request by efritz**
This janitor will periodically ensure that a bundle file exists for every processed upload record in the database. If there is a missing bundle file, the associated db record is deleted.

The bundle manager has a symmetric janitor process that will delete any bundle files for which there is no record.

This should keep the upload record and the processed files in sync in case there is any manual intervention on the filesystem, or programming errors that create spurious files (like opening empty SQLite dbs, as we did for a short period of time on dot-com).

### CODE REVIEW

**Comment by efritz** [comment]
> could this process be in the bundle manager to avoid network hops?

The bundle manager does not currently talk to the database (by design). This is not in a critical path (it only runs twice an hour) and requests are batched to reduce this. We already do the same thing in the reverse direction.

> if this is going to be a periodic process, what's the advantage of this approach vs letting them fall out of the db automatically when they get stale?

Nothing falls out of the db based on age alone. The current eviction is in the bundle manager janitor: while there's not enough disk space, it asks the api-server to delete the oldest upload record and return its id, then the bundle manager deletes the corresponding bundle file.

> does the process for deleting dumps from the db because they're on old commits gracefully handle the bundle already not existing?

Yes, the bundle manager's janitor will request a db record be deleted, it will return it and the bundle manager will try to delete the bundle that just happens not to exist. Right now a stat fails, but I'm pushing a commit to this Pr to fix that condition (good catch!).

> i'm worried about the full table scan of lsif_dumps, could we add a note/file an issue that that's going to be a problem as we scale so we don't forget about it?

I added the `order by` clause on that query specifically to force an index scan (without it it does do a sequential scan on the table). It should be efficient, especially for a query that doesn't happen on any critical path. If it becomes inefficient we can easily change this to do limit/offset or keyset pagination with the same general structure.

**Comment by efritz** [comment]
Also, in general, I think it's good to have reconciliation/cleanup processes like these in place, even if in normal operation things should be cleaning up after themselves. There are too many failure modes in software to be able to say for sure that state will remain consistent when there are 2+ processes involved.

File: cmd/precise-code-intel-api-server/internal/janitor/janitor.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// for more specifics: removeProcessedUploadsWithoutBundleFile.
+func (j *Janitor) Run() {
+	for {
+		if err := j.run(); err != nil {
+			j.metrics.Errors.Inc()
+			log15.Error("Failed to run janitor process", "err", err)
+		}
+
+		select {
+		case <-time.After(j.janitorInterval):
```
Comment (uwedeportivo) [inline cmd/precise-code-intel-api-server/internal/janitor/janitor.go:47]: this is totally fine, but if you want you can create a time.Ticker outside the for loop and keep using it. you can also close it when you're done

Comment (efritz) [reply cmd/precise-code-intel-api-server/internal/janitor/janitor.go:47]: A ticker will fire regardless of how long the inner function takes, so if you have one long-running action you may run them back to back. With after you wait at least the interval between invocations, which seems alright for a cleanup process.

Comment (uwedeportivo) [reply cmd/precise-code-intel-api-server/internal/janitor/janitor.go:47]: i see, makes sense. sounds good to me

**Pull request merged**